src/Features/Core/Portable/Diagnostics/Service/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs

```diff
@@ -16,14 +18,45 @@ namespace Microsoft.CodeAnalysis.Diagnostics;
 
 internal sealed partial class DiagnosticAnalyzerService
 {
+    private sealed class ChecksumAndAnalyzersEqualityComparer
+        : IEqualityComparer<(Checksum checksum, ImmutableArray<DiagnosticAnalyzer> analyzers)>
+    {
+        public static readonly ChecksumAndAnalyzersEqualityComparer Instance = new();
+
+        public bool Equals((Checksum checksum, ImmutableArray<DiagnosticAnalyzer> analyzers) x, (Checksum checksum, ImmutableArray<DiagnosticAnalyzer> analyzers) y)
+        {
+            if (x.checksum != y.checksum)
+                return false;
+
+            // Fast path for when the analyzers are the same reference.
+            return x.analyzers == y.analyzers || x.analyzers.SetEquals(y.analyzers);
+        }
+
+        public int GetHashCode((Checksum checksum, ImmutableArray<DiagnosticAnalyzer> analyzers) obj)
+        {
+            var hashCode = obj.checksum.GetHashCode();
+
+            // Use addition so that we're resilient to any order for the analyzers.
+            foreach (var analyzer in obj.analyzers)
+                hashCode += analyzer.GetHashCode();
+
+            return hashCode;
+        }
+    }
+
     /// <summary>
-    /// Cached data from a <see cref="ProjectState"/> to the last <see cref="CompilationWithAnalyzersPair"/> instance
-    /// created for it.  Note: the CompilationWithAnalyzersPair instance is dependent on the set of <see
-    /// cref="DiagnosticAnalyzer"/>s passed along with the project.  As such, we might not be able to use a prior cached
-    /// value if the set of analyzers changes.  In that case, a new instance will be created and will be cached for the
-    /// next caller.
+    /// Cached data from a <see cref="ProjectState"/> to the <see cref="CompilationWithAnalyzersPair"/>s
+    /// we've created for it.  Note: the CompilationWithAnalyzersPair instance is dependent on the set of <see
+    /// cref="DiagnosticAnalyzer"/>s passed along with the project.
+    /// <para/>
+    /// The value of the table is a <see cref="ConcurrentDictionary{TKey, TValue}"/> that maps from the 
+    /// <see cref="Project"/> checksum the set of <see cref="DiagnosticAnalyzer"/>s being requested.
     /// </summary>
-    private static readonly ConditionalWeakTable<ProjectState, StrongBox<(Checksum checksum, ImmutableArray<DiagnosticAnalyzer> analyzers, CompilationWithAnalyzersPair? compilationWithAnalyzersPair)>> s_projectToCompilationWithAnalyzers = new();
+    private static readonly ConditionalWeakTable<
+        ProjectState,
+        ConcurrentDictionary<
+            (Checksum checksum, ImmutableArray<DiagnosticAnalyzer> analyzers),
+            AsyncLazy<CompilationWithAnalyzersPair?>>> s_projectToCompilationWithAnalyzers = new();
```

Comment:
NOte: i designed it this way so that we dont' toss out a very good 'full compilation with analyzers' and replace it with a 'compilation with analyzers with a single analyzer in it'.